### DISCUSSION TITLE
Making python bindings for a project with CUDA kernels.

### DISCUSSION THREAD

**Discussion by pmichels19** [wjakob/nanobind#401] [Q&A] [answered]
Hi everyone! I am currently working on a project that requires me to build python bindings. For performance reasons I decided that nanobind would be a great option for this. However, I am running into a bit of an issue with compiling the CUDA kernels that are included in the project.

My CMakeLists.txt looks like this:
```cmake
cmake_minimum_required(VERSION 3.15...3.27)
project(my_project LANGUAGES CXX CUDA)
enable_language(CUDA)

# Warn if the user invokes CMake directly
if (NOT SKBUILD)
  message(WARNING <standard warning is here>)
endif()

# Try to import all Python components potentially needed by nanobind
find_package(Python 3.11
    REQUIRED COMPONENTS Interpreter Development.Module
    OPTIONAL_COMPONENTS Development.SABIModule
)
find_package(nanobind CONFIG REQUIRED)
find_package(CUDAToolkit REQUIRED)

set(MY_PROJECT_SOURCES
    # Source files go here
    src/pch.cpp
    src/types.cpp
    src/logging.cpp
    src/display.cpp
    src/main.cpp
)

set(MY_PROJECT_INCLUDES
    # Include folders go here
    Dependencies/x64/argparse/include
    Dependencies/x64/CImg/include
    Dependencies/x64/cuFFTDx/include
    Dependencies/x64/fftw/include
    Dependencies/x64/glm/include
    Dependencies/x64/libpng/include
    Dependencies/x64/zlib/include
    ${CUDAToolkit_INCLUDE_DIRS}
)

set(MY_PROJECT_DEPENDENCIES
    "${CMAKE_SOURCE_DIR}/Dependencies/x64/fftw/lib/libfftw3-3.lib"
    "${CMAKE_SOURCE_DIR}/Dependencies/x64/fftw/lib/libfftw3f-3.lib"
    "${CMAKE_SOURCE_DIR}/Dependencies/x64/libpng/lib/libpng16.lib"
    "${CMAKE_SOURCE_DIR}/Dependencies/x64/libpng/lib/libpng16_static.lib"
    "${CMAKE_SOURCE_DIR}/Dependencies/x64/zlib/lib/zlib.lib"
    "${CMAKE_SOURCE_DIR}/Dependencies/x64/zlib/lib/zlibstatic.lib"
    CUDA::cufft
    CUDA::cudart
    #${CUDAToolkit_LIBRARY_DIR}
)

add_library(MY_PROJECT_CUDA_KERNELS STATIC my-project/src/kernels.cu)
set_target_properties(MY_PROJECT_CUDA_KERNELS
    PROPERTIES
    POSITION_INDEPENDENT_CODE ON
    CUDA_VISIBILITY_PRESET "hidden"
    CUDA_SEPARABLE_COMPILATION ON
)

target_link_libraries(MY_PROJECT_CUDA_KERNELS PRIVATE ${MY_PROJECT_DEPENDENCIES} ${Python_LIBRARIES})
target_include_directories(MY_PROJECT_CUDA_KERNELS PRIVATE ${MY_PROJECT_INCLUDES})

add_library(MY_PROJECT_LIBRARY STATIC ${MY_PROJECT_SOURCES})
set_target_properties(MY_PROJECT_LIBRARY
    PROPERTIES
    CXX_STANDARD 17
    POSITION_INDEPENDENT_CODE ON
    CUDA_VISIBILITY_PRESET "hidden"
    CUDA_SEPARABLE_COMPILATION ON
)

target_link_libraries(MY_PROJECT_LIBRARY PRIVATE ${MY_PROJECT_DEPENDENCIES} ${MY_PROJECT_CUDA_KERNELS})
target_include_directories(MY_PROJECT_LIBRARY PRIVATE ${MY_PROJECT_INCLUDES})

nanobind_add_module(my_project_ext
    STABLE_ABI
    NB_STATIC
    src/my_project_ext.cpp
    ${MY_PROJECT_SOURCES}
    src/kernels.cu
)

target_link_libraries(my_project_ext PRIVATE ${MY_PROJECT_DEPENDENCIES} ${MY_PROJECT_LIBRARY})
target_include_directories(my_project _ext PRIVATE ${MY_PROJECT_INCLUDES})

install(TARGETS my_project_ext LIBRARY DESTINATION my_project)
```

In `my_project_ext.cpp` I have a very generic test function as well as a constructor for the main class:
```cpp
#include "pch.h"

#include <nanobind/nanobind.h>

namespace nb = nanobind;
using namespace nb::literals;

int add( int i1, int i2 ) {
    return i1 + i2;
}

NB_MODULE(my_project_ext, m) {
    m.def( "add", &add, "a"_a, "b"_a = 1,
           "This function adds two numbers and increments if only one is provided." );

    nb::class_<Main>(m, "Main").def(nb::init<int, int, int, const std::string&>());
}
```

Unfortunately, I am getting quite an extensive error that I am not sure how to fix:
```
"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.3\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\bin\HostX64\x64" -x cu   -I"C:\Users\<user>\Documents\my-project\Dependencies\x64\argparse\include" -I"C:\Users\<user>\Documents\my-project\Dependencies\x64\CImg\include" -I"C:\Users\<user>\Documents\my-project\Dependencies\x64\cuFFTDx\include" -I"C:\Users\<user>\Documents\my-project\Dependencies\x64\fftw\include" -I"C:\Users\<user>\Documents\my-project\Dependencies\x64\glm\include" -I"C:\Users\<user>\Documents\my-project\Dependencies\x64\libpng\include" -I"C:\Users\<user>\Documents\my-project\Dependencies\x64\zlib\include" -I"C:\Users\<user>\AppData\Local\Programs\Python\Python311\Lib\site-packages\nanobind\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.3\include" -IC:\Users\<user>\AppData\Local\Programs\Python\Python311\Include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.3\include"     --keep-dir x64\Release  -maxrregcount=0   --machine 64 --compile -cudart static -std=c++17 --generate-code=arch=compute_52,code=[compute_52,sm_52] /bigobj /MP /Os -Xcompiler="/EHsc -Ob2"   -D_WINDOWS -DNDEBUG -D"CMAKE_INTDIR=\"Release\"" -Dmy_project_ext_EXPORTS -D_WINDLL -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG -D"CMAKE_INTDIR=\"Release\"" -Dmy_project_ext_EXPORTS -Xcompiler "/EHsc /W1 /nologo /O2 /FS   /MD " -Xcompiler "/Fdmy_project_ext.dir\Release\vc143.pdb" -o my_project_ext.dir\Release\kernels.obj "C:\Users\<user>\Documents\my-project\src\kernels.cu"
        nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
```

Is there anyone who has made python bindings with a CUDA kernel before who is capable of providing some insight or maybe sharing how they did it? I have been stuck on this problem for...longer than I would like to admit really 😅

**Comment by wjakob** [comment]
Could it be something like your projected containing spaces, and one of the include directives not having quotes?

**Reply by pmichels19** [reply]
Mmh, I had missed that `C:\Users\<user>\AppData\Local\Programs\Python\Python311\Include` does not have quotes around it but I am uncertain how I should fix that. Particularly because I am not manually including that. I assume that is something that nanobind does? As for the spaces in the path, I already checked and that did not fix the issue sadly.

**Reply by wjakob** [reply]
Could you check if this patch to nanobind fixes the issue on your end?

```diff
diff --git a/cmake/nanobind-config.cmake b/cmake/nanobind-config.cmake
index cb6cfb0..36611b2 100644
--- a/cmake/nanobind-config.cmake
+++ b/cmake/nanobind-config.cmake
@@ -214,12 +214,12 @@ function (nanobind_build_library TARGET_NAME)
     target_link_libraries(${TARGET_NAME} PRIVATE tsl::robin_map)
   else()
     target_include_directories(${TARGET_NAME} PRIVATE
-      ${NB_DIR}/ext/robin_map/include)
+      "${NB_DIR}/ext/robin_map/include")
   endif()

   target_include_directories(${TARGET_NAME} PUBLIC
-    ${Python_INCLUDE_DIRS}
-    ${NB_DIR}/include)
+    "${Python_INCLUDE_DIRS}"
+    "${NB_DIR}/include")

   target_compile_features(${TARGET_NAME} PUBLIC cxx_std_17)
   nanobind_set_visibility(${TARGET_NAME})
```

**Reply by pmichels19** [reply]
Strangely it doesn't seem to change anything. I'd expect quotes to appear in the command that is run 🤔 I already confirmed that no cached version is used and it is actually taking from my site-packages folder.

**Reply by wjakob** [reply]
If in doubt, you can add a `message(STATUS "..")` to `nanobind-config.cmake` just to see that this is run. Is there any chance that this problematic line is an include directive from your project and not nanobind?

**Reply by pmichels19** [reply]
Yeah I checked using a message like that 😅 should have clarified a bit on that. The other question though: very certain. If I comment out everything from the `nanobind_add_module` call and below I don't get any errors.

**Reply by pmichels19** [reply]
An I think related issue has now popped up (with @matthewdcong 's solution and the quotes around includes both applied) where the module compiles but when I try to import it I get the following error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python311\Lib\site-packages\my_project\__init__.py", line 1, in <module>
    from .my_project_ext import add
ImportError: DLL load failed while importing my_project_ext: The specified module could not be found.
```

I already checked with [Dependencies](https://github.com/lucasg/Dependencies) to see if all the dlls that are needed are there and that all seems to be fine. Is this an issue that has been encountered before? Because I am not sure how to fix this 🤔

**Reply by wjakob** [reply]
Were the quotes in my patch actually ultimately needed? If so, all of them or a specific one? I would be quite surprised if the contents of `Python_INCLUDE_DIRS` aren't themselves quoted already given that it indicates a plurality of paths, and wrapping those into quotes hardly makes sense.

**Reply by matthewdcong** [reply]
I think this DLL issue might be due to a version mismatch/library conflict that's independent of nanobind. I'd start with a simple main.cu file with the `add` Python binding example similar to what you've shown above and slowly add the required libraries back in. Note that the .cu suffix is intentional here, as it will prompt CMake to compile with nvcc instead of msvc (the former being a valid host compiler as well).

**Reply by pmichels19** [reply]
Alright, I'll give that a go! As for the quotes: it doesn't actually seem to matter 😅 still thanks for the quick responses though!

**Reply by pmichels19** [reply]
After some trying I've been able to at least nail down the problem a little more: it seems to be the cufft library.
My `my_project_ext.cpp` looks like this at the moment:
```c++
#include <cufft.h>

#include <nanobind/nanobind.h>

namespace nb = nanobind;

using namespace nb::literals;

int add( int i1, int i2 ) {
    cufftPlan3d(nullptr, 0, 0, 0, CUFFT_C2C);
    return i1 + i2;
}

NB_MODULE(my_project_ext, m) {
    m.def( "add", &add, "a"_a, "b"_a = 1,
           "This function adds two numbers and increments if only one is provided." );
}
```
With a CMakeLists.txt like this:
```
cmake_minimum_required(VERSION 3.15...3.27)
project(my_project LANGUAGES CXX CUDA)
enable_language(CUDA)

# Warn if the user invokes CMake directly
if (NOT SKBUILD)
    message(WARNING <standard warning is here>)
endif()

# Try to import all Python components potentially needed by nanobind
find_package(Python 3.11 REQUIRED COMPONENTS Interpreter Development.Module OPTIONAL_COMPONENTS Development.SABIModule)
find_package(nanobind CONFIG REQUIRED)
find_package(CUDAToolkit REQUIRED)

set(MY_PROJECT_INCLUDES
    # Include folders go here
    ${CUDAToolkit_INCLUDE_DIRS}
)

set(MY_PROJECT_DEPENDENCIES
    "${CUDAToolkit_LIBRARY_DIR}/cufft.lib"
)

nanobind_add_module(my_project_ext
    STABLE_ABI
    NB_STATIC
    src/my_project_ext.cpp
)

target_link_libraries(my_project_ext PRIVATE ${MY_PROJECT_DEPENDENCIES})
target_include_directories(my_project_ext PRIVATE ${MY_PROJECT_INCLUDES})

install(TARGETS my_project_ext LIBRARY DESTINATION my_project)
```
This still gives the same DLL import error.

I don't know if this is still an issue that should be kept on this GitHub by the way. I don't know if it is maybe a good idea to post this issue on the NVIDIA rather than here.

**Reply by wjakob** [reply]
I've read at some point there is some way to get DLL loading status messages on windows that could help to further pinpoint the issue. But I don't remember the details, perhaps someone else here knows?

**Reply by wjakob** [reply, marked as answer]
You need Process Monitor: https://learn.microsoft.com/en-us/sysinternals/downloads/procmon

**Reply by pmichels19** [reply]
Ah that actually got me a lot further! Apparently it was looking for the msvcp140, cufft and cudart dlls and failing to find them in the folders where it was looking. I'll see if I can fix that somehow, thanks for the fast replies and the help!

**Comment by matthewdcong** [comment]
`nvcc` doesn't support the `/bigobj`, `/MP`/, and `/Os` compiler options and is interpreting those as paths. Are you on the latest ToT master nanobind? PR #383 was sufficient to fix the issue for me.

**Reply by pmichels19** [reply]
Ah that has at least fixed the compiling issue! I do get strange behavior now when trying to import the module 🤔 seems like the CUDA libraries might not be linked properly (I think? Looking at the dependencies the cudart has an incorrect checksum and cufft isn't even there...)
